website/www/site/content/en/documentation/ml/online-clustering.md

```diff
@@ -0,0 +1,211 @@
+5. Normalizing the embedding for Clustering
+6. Performing BIRCH Clustering using Stateful Processing
+7. Printing the texts assigned to clusters
+
+The code snippet for first two steps of pipeline where message from PubSub is read and converted into a dictionary
+
+{{< highlight >}}
+    docs = (
+        pipeline
+        | "Read from PubSub"
+        >> ReadFromPubSub(subscription=cfg.SUBSCRIPTION_ID, with_attributes=True)
+        | "Decode PubSubMessage" >> beam.ParDo(Decode())
+    )
+{{< /highlight >}}
+
+
+We now closely look at three important steps of pipeline where we tokenize the text, fed the tokenized text to get embedding from a Transformer based Language Model and performing clustering using [Stateful Processing](https://beam.apache.org/blog/stateful-processing/).
+
+
+### Getting Embedding from a Language Model
+
+In order to do clustering with text data, we first need to map the text into vectors of numerical values suitable for statistical analysis. We use a transformer based language model called [sentence-transformers/stsb-distilbert-base/stsb-distilbert-base](https://huggingface.co/sentence-transformers/stsb-distilbert-base). It maps sentences & paragraphs to a 768 dimensional dense vector space and can be used for tasks like clustering or semantic search. But, we first need to tokenize the text as the language model is expecting a tokenized input instead of raw text.
+
+Tokenization can be seen as a preprocessing task as it transforms text in a way that it can be fed into the model for getting predictions.
+
+{{< highlight >}}
+    normalized_embedding = (
+        docs
+        | "Tokenize Text" >> beam.Map(tokenize_sentence)
+{{< /highlight >}}
+
+Here, `tokenize_sentence` is a function that takes a dictionary with a text and an id, tokenizes the text, and returns a tuple (text, id) and the tokenized output.
+
+
+Tokenized output is then passed to the language model for getting the embeddings. For getting embeddings from language model, we use `RunInference()` from beam.
+
+{{< highlight >}}
+    | "Get Embedding" >> RunInference(KeyedModelHandler(model_handler))
+
+{{< /highlight >}}
+
+After getting the embedding for each twitter text, the embeddings are normalized as it helps to make better clusters.
+
+{{< highlight >}}
+    | "Normalize Embedding" >> beam.ParDo(NormalizeEmbedding())
+
+{{< /highlight >}}
+
+
+### StatefulOnlineClustering
+As the data is coming in a streaming fashion, so to cluster them we need an iterative clustering algorithm like BIRCH. As, the algorithm is iterative, we need a mechanism to store the previous state so that when a twitter text arrives, it can be updated accordingly.  ** Stateful Processing ** enables a `DoFn` with the ability to have a persistent state which can be read and written during the processing of each element. One can read about Stateful Processing in the official documentation from Beam: [Link](https://beam.apache.org/blog/stateful-processing/).
```

Comment:
```suggestion
As the data is coming in a streaming fashion, so to cluster them we need an iterative clustering algorithm like BIRCH. As, the algorithm is iterative, we need a mechanism to store the previous state so that when a twitter text arrives, it can be updated accordingly.  *Stateful Processing* enables a `DoFn` to have persistent state which can be read and written during the processing of each element. One can read about Stateful Processing in the official documentation from Beam: [Link](https://beam.apache.org/blog/stateful-processing/).
```